File: DirectProgramming/C++/Jupyter/OpenMP-offload-training/datatransfer/main_data_region.cpp
//==============================================================
// Copyright © 2020 Intel Corporation
//
// SPDX-License-Identifier: MIT
// =============================================================

#include <omp.h>
#include <stdio.h>

constexpr int ARRAY_SIZE = 256;

void init1(float *x, int N) {
  for (int i = 0; i < N; i++) x[i] = 1.0;
}
void init2(float *x, int N) {
  for (int i = 0; i < N; i++) x[i] = 2.0;
}
int main() {
  float x[ARRAY_SIZE], y[ARRAY_SIZE];
  double tb, te;
  int correct_count = 0;

  init1(x, ARRAY_SIZE);
  init1(y, ARRAY_SIZE);

  printf("Number of OpenMP Devices: %d\n", omp_get_num_devices());

  tb = omp_get_wtime();

#include "lab/target_data_region.cpp"

  te = omp_get_wtime();

  printf("Time of kernel: %lf seconds\n", te - tb);

  for (int i = 0; i < ARRAY_SIZE; i++)
    if (x[i] == 4.0) correct_count++;

  printf("Test: %s\n", (correct_count == ARRAY_SIZE) ? "PASSED!" : "Failed");
}

File: DirectProgramming/Fortran/Jupyter/OpenMP_Offload_Fortran/datatransfer/main_data_region.f90
!==============================================================
! Copyright © 2020 Intel Corporation
!
! SPDX-License-Identifier: MIT
! =============================================================

subroutine init (x, n, val)
        integer, intent(in) :: n
        real(kind=4), intent(out), dimension(n) :: x
        real(kind=4), intent(in) :: val
        do i=1,n
                x(i)=val
        end do
end subroutine init

program main
        use omp_lib
        integer, parameter :: ARRAY_SIZE=256
        integer :: i, correct_count=0;
        real(kind=4) :: x(ARRAY_SIZE), y(ARRAY_SIZE) 
        real(kind=4) :: a = 1.5, tolerance=0.01
        double precision :: tb, te

        call init (x, ARRAY_SIZE, 1.0)
        call init (y, ARRAY_SIZE, 1.0)

        print *, "Number of OpenMP Devices: ", omp_get_num_devices()

        tb = omp_get_wtime() 

        ! Written by Jupyter Notebook
        INCLUDE 'lab/target_data_region.f90'

        te = omp_get_wtime() 

        print '("Work took ",(f7.5)," seconds.")', te-tb

        !Check Results
        do i=1, ARRAY_SIZE
                if (x(i)==4.0 ) correct_count=correct_count+1
        end do
        if (correct_count == ARRAY_SIZE) then
                print *, "Test: Passed"
        else
                print *, "Test: Failed"
        end if
end program main
